### DISCUSSION TITLE
Query vs Javascript array methods performance

### DISCUSSION THREAD

**Discussion by KrisLau** [Nozbe/WatermelonDB#1395] [General]
When having to choose between doing a query and an array method (`.filter()`, `.map()`, etc) or lodash stuff like `_.find()`, which one should I use?

**Comment by scarlac** [comment]
@KrisLau You should generally never use lodash's _.find, _.filter, and _.map. Use the native equivalent: [].find, [].filter, and [].map() for maximum speed.
Regarding using SQL to process data compared to processing data from a query yourself, it's likely that SQL is faster than your code, simply because it's native, but also because it can analyze the dataset for you.
In some some situations you have an expectation about the dataset size that SQLite doesn't have and in those situations it is better to "download" the data first, then process it in JS.

**Comment by ororsatti** [comment]
Generally, using the built in orm is better since it runs on the native side,
which doesnt block your thread, and simply runs faster.
But, taking it a step further, I dont think on complex queries, where you cant simply use wm orm,
I would preform the action on JS, since its easier to maintain in the long run.
